File: add_two_numbers.java


// Definition for singly-linked list.

// import java.util.ArrayList;
// import java.util.List;

class ListNode
{
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}



class Solution
{
    public ListNode addTwoNumbers(ListNode l1, ListNode l2)
    {
        ListNode temp = new ListNode(0);
        ListNode change_temp = temp;
        int overflow = 0;

        while (l1 != null || l2 != null || overflow != 0)
        {
            int num1 = (l1 != null) ? l1.val : 0;
            int num2 = (l2 != null) ? l2.val : 0;
            int sum = num1 + num2 + overflow;

            change_temp.next = new ListNode(sum % 10);
            change_temp = change_temp.next;
            overflow = sum / 10;

            l1 = (l1 != null) ? l1.next : null;
            l2 = (l2 != null) ? l2.next : null;
        }

        return temp.next;
    }
}

File: python/medium/add-two-numbers.py
# medium (https://leetcode.com/problems/add-two-numbers/)


from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        temp = ListNode(0)
        change_temp = temp
        overflow = 0

        while l1 or l2 or overflow:
            num1 = l1.val if l1 else 0
            num2 = l2.val if l2 else 0

            total = num1 + num2 + overflow
            num = total % 10
            overflow = total // 10
    
            new_node = ListNode(num)
            change_temp.next = new_node
            change_temp = change_temp.next

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return temp.next